Make pipeWrite and pipeRead noexcept

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45783 Make pipeWrte and pipeRead noexcept**

After the previous device maps commits, `pipeWrite` might throw. In
this case, if we increment active calls before `pipeWrite` on the
caller, that active call won't be decremented properly when `pipeWrite`
throws. As a result, `shutdown` can silently timeout. I noticed this
as some tests take more than 60s to finish.

This commit extract the tensor device checking logic out of pipeWrite,
and make sure the error is thrown before the active call count is
incremented.

Differential Revision: [D24094803](https://our.internmc.facebook.com/intern/diff/D24094803)

mrshenli committed:
Only increase TensorPipe agent client active calls when pipeWrite succeeds

[ghstack-poisoned]

mrshenli committed:
Update on "Only increase TensorPipe agent client active calls when pipeWrite succeeds"




[ghstack-poisoned]

mrshenli committed:
Update on "Only increase TensorPipe agent client active calls when pipeWrite succeeds"




[ghstack-poisoned]

mrshenli committed:
Update on "Only increase TensorPipe agent client active calls when pipeWrite succeeds"


As `pipeWrite` might throw, if increase active calls before `pipeWrite`,
that active call won't be decremented properly when `pipeWrite` throws.
As a result, `shutdown` can silently timeout. I noticed this as some
tests take more than 60s to finish.

[ghstack-poisoned]

lw commented:
Good catch, and thanks for fixing it!

I am worried about one thing: in principle it should be possible to have a race condition where the TensorPipe event loops invoke the callbacks before the main thread actually gets to the point of increasing the active client count. Since the callbacks decrease that count, we may get to a situation where, for a brief moment, the count may be negative. I don't know if this can cause issues in practice, but perhaps we should try to avoid it?

Apart from adding a try/catch block, the only other idea I have to solve this may actually be over-engineered: we could create a small movable-only class, which increases the count upon construction, and decreases it upon destruction. We create such an object before calling pipeWrite, and capture it in the lambda that we pass to pipeWrite. This means that if pipeWrite succeeds, the object will stay alive until the callback is called. But if it fails, the callback will be destroyed and, with it, the object, and thus the counter will be decreased.

mrshenli commented:
> Apart from adding a try/catch block, the only other idea I have to solve this may actually be over-engineered: we could create a small movable-only class, which increases the count upon construction, and decreases it upon destruction. We create such an object before calling pipeWrite, and capture it in the lambda that we pass to pipeWrite. This means that if pipeWrite succeeds, the object will stay alive until the callback is called. But if it fails, the callback will be destroyed and, with it, the object, and thus the counter will be decreased.

Thanks for the suggestion! I like this solution. Let me add that.

lw commented:
Uhm, in fact there may be an even simpler one: instead of throwing inside pipeWrite, we can perform the same checks _outside_ of it, and throw in case they fail, and only once we've made sure they all pass we can increase the count. (For extra safety we could also re-do those checks inside pipeWrite, but at that point raise an assert as it indicates a programming error and not a user error)

mrshenli committed:
Update on "Only increase TensorPipe agent client active calls when pipeWrite succeeds"


As `pipeWrite` might throw, if increase active calls before `pipeWrite`,
that active call won't be decremented properly when `pipeWrite` throws.
As a result, `shutdown` can silently timeout. I noticed this as some
tests take more than 60s to finish.

Differential Revision: [D24094803](https://our.internmc.facebook.com/intern/diff/D24094803)

[ghstack-poisoned]

mrshenli commented:
> Uhm, in fact there may be an even simpler one: instead of throwing inside pipeWrite, we can perform the same checks outside of it, and throw in case they fail, and only once we've made sure they all pass we can increase the count. (For extra safety we could also re-do those checks inside pipeWrite, but at that point raise an assert as it indicates a programming error and not a user error)

Are we sure that `pipe->write` will never throw?

mrshenli committed:
Update on "Only increase TensorPipe agent client active calls when pipeWrite succeeds"


As `pipeWrite` might throw, if increase active calls before `pipeWrite`,
that active call won't be decremented properly when `pipeWrite` throws.
As a result, `shutdown` can silently timeout. I noticed this as some
tests take more than 60s to finish.

Differential Revision: [D24094803](https://our.internmc.facebook.com/intern/diff/D24094803)

[ghstack-poisoned]

lw commented:
> Are we sure that pipe->write will never throw?

Yes, that's a guarantee provided by TensorPipe, which doesn't use exceptions at all internally, and instead uses explicit error return values (or arguments to callbacks). The only case where anything might throw an exception is when an assert is violated, but those are unrecoverable programming errors and thus we shouldn't try to catch and handle them.

mrshenli committed:
Update on "Only increase TensorPipe agent client active calls when pipeWrite succeeds"


As `pipeWrite` might throw, if increase active calls before `pipeWrite`,
that active call won't be decremented properly when `pipeWrite` throws.
As a result, `shutdown` can silently timeout. I noticed this as some
tests take more than 60s to finish.

Differential Revision: [D24094803](https://our.internmc.facebook.com/intern/diff/D24094803)

[ghstack-poisoned]

mrshenli reviewed (commented):
Updated. Request and response now share the same error checking process. Request would let it fail, and response will try-catch that and convert it to the message. Let me know if this looks OK to you. Thanks! @lw

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -1075,14 +1025,58 @@ void TensorPipeAgent::markFutureWithError(
+  const auto& iter = deviceMaps.find(remoteName);
+  if (iter == deviceMaps.end()) {
+    for (const auto& t : message.tensors()) {
+      TORCH_CHECK(
+        t.device().is_cpu(),
+        "TensorPipe RPC backend only supports CPU tensors by default, please "
+        "move your tensors to CPU before sending them over RPC, or call "
+        "`set_device_map` on `TensorPipeRpcBackendOptions` to explicitly "
+        "configure device mapping. ",
+        messageType,
+        " device mapping is not available for destination ",
+        remoteName,
+        ", but found tensor on device: ",
+        t.device());
+    }
+    return {};
+  } else {
+    std::vector<c10::DeviceIndex> deviceIndices;
+    deviceIndices.reserve(message.tensors().size());
+    const auto& deviceMap = iter->second;
+    for (const auto& t : message.tensors()) {
+      const auto deviceIter = deviceMap.find(t.device().index());
+      if (deviceIter == deviceMap.end()) {
+        TORCH_CHECK(
+          t.device().is_cpu(),
+          "TensorPipe RPC backend only supports CPU tensors by default. ",
+          messageType,
+          " device mapping is not available for destination ",
+          remoteName,
+          " for device ",
+          t.device(),
+          " but received a tensor on that device.");
```
There are two nearly-identical instances of `TORCH_CHECK(t.device().is_cpu(), ...)` and there is also the `checkCPUTensor` function that does the same thing. Is this last function even used anymore? Could we consolidate this?

mrshenli replied:
yes, should have removed `checkCPUTensor`.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -1075,14 +1025,58 @@ void TensorPipeAgent::markFutureWithError(
-      // The future's callbacks may schedule further RPCs, increasing the count.
-      // Thus we must decrease it after completing the future, otherwise it may
-      // briefly dip to zero and trick join into thinking all work is done.
-      decreaseCallCount(clientActiveCalls_);
```
It seems that `markFutureWithError` is also called when a timeout expires, which means we used to decrease the count when that happened, but now we don't do it anymore. Was that intended behavior? If so, we should try to preserve it...

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -703,10 +644,23 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
+  auto devices = getDevicesForTensors(
+      clientPipe.pipe_->getRemoteName(), requestMessage);
+
+  // Increase clientActiveCalls_ and pass it to pipeWrite and pipeRead
+  // callbacks. It will decrease clientActiveCalls_ accordingly on destruction.
+  auto clientCallCounter =
+      std::make_shared<CallCounter>(*this, clientActiveCalls_);
```
Could we move it to the same exact spot where we did the increase before? I am a bit worried about a race condition with the timeout thread, which could fire and decrease the count before we reach this point (see also my comment further down about the timeout thread currently not decreasing the count).

Also, it is a shame that we need to wrap this in a shared_ptr and thus perform an extra heap allocation just for this small thing. Unfortunately I suspect it's necessary because even if we made the CallCounter a full RAII object, it would be movable-only, and we cannot capture movable-only objects in std::functions. (On the TensorPipe side we are thinking of rolling our own non-copyable function wrapper which we could then use here too...)

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -703,10 +644,23 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
   VLOG(1) << "RPC agent for " << workerInfo_.name_ << " is sending request #"
           << messageId << " to " << clientPipe.pipe_->getRemoteName();
 
+  auto devices = getDevicesForTensors(
+      clientPipe.pipe_->getRemoteName(), requestMessage);
```
Shouldn't we do this even before we add the request to the timeout queue?

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -344,6 +349,21 @@ class TensorPipeAgent : public RpcAgent {
   std::mutex networkDataMutex_;
 
+  struct CallCounter final {
+    CallCounter(TensorPipeAgent& agent, int32_t& counter)
+        : agent_(agent), counter_(counter) {
+      agent_.increaseCallCount(counter_);
+    }
+
+    ~CallCounter() {
+      agent_.decreaseCallCount(counter_);
+    }
```
When defining a custom constructor and destructor, the compiler will still add an implicitly-declared copy constructor. Right now this doesn't cause problems, because we allocate the CallCounter in a shared_ptr and never copy it. However, keeping that copy constructor around introduces the risk that at some point someone will implicitly call it by accident and that will lead to hard-to-debug issues: the copy will not increase the count (because it will use a different constructor) however it will still decrease it (as it uses the same destructor). I'd suggest we explicitly delete the copy constructor and assignment operator, to be safe.

lw reviewed (commented):
I think there may be some issues remaining, I left some inline comments.

One way to address both the issue of the extra heap allocation and the issue of timeouts not decreasing the count _could_ be to add the CallCounter as a field of the AtomicFutureMessage struct.

However, that made me realize that there may be another simpler solution to this problem, which would avoid this RAII-like class: we could increase the call count like we did before, and then _install a callback on the future_ that decreases it. This will ensure that there is a single piece of code that is responsible for decreasing it, and that it will always be executed no matter who marks the future as complete and no matter whether it does so with a success or an error.

mrshenli committed:
Update on "Make pipeWrte and pipeRead noexcept"


After the previous device maps commits, `pipeWrite` might throw. In
this case, if we increment active calls before `pipeWrite` on the
caller, that active call won't be decremented properly when `pipeWrite`
throws. As a result, `shutdown` can silently timeout. I noticed this
as some tests take more than 60s to finish.

This commit extract the tensor device checking logic out of pipeWrite,
and make sure the error is thrown before the active call count is
incremented.

Differential Revision: [D24094803](https://our.internmc.facebook.com/intern/diff/D24094803)

[ghstack-poisoned]

mrshenli commented:
>  we could increase the call count like we did before, and then install a callback on the future that decreases it. This will ensure that there is a single piece of code that is responsible for decreasing it, and that it will always be executed no matter who marks the future as complete and no matter whether it does so with a success or an error.

This looks great. I removed the `CallCounter` class and reverted to the previous increment/decrement logic. So the only change made by this PR now is to throw before we increment the call count. Shall we get this to master first to speed up CIs? I can work on the above proposal as a followup PR.

mrshenli committed:
Update on "Make pipeWrte and pipeRead noexcept"


After the previous device maps commits, `pipeWrite` might throw. In
this case, if we increment active calls before `pipeWrite` on the
caller, that active call won't be decremented properly when `pipeWrite`
throws. As a result, `shutdown` can silently timeout. I noticed this
as some tests take more than 60s to finish.

This commit extract the tensor device checking logic out of pipeWrite,
and make sure the error is thrown before the active call count is
incremented.

Differential Revision: [D24094803](https://our.internmc.facebook.com/intern/diff/D24094803)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -1083,6 +1020,53 @@ void TensorPipeAgent::markFutureWithError(
+    deviceIndices.reserve(message.tensors().size());
+    const auto& deviceMap = iter->second;
+    for (const auto& t : message.tensors()) {
+      const auto deviceIter = deviceMap.find(t.device().index());
+      if (deviceIter == deviceMap.end()) {
+        TORCH_CHECK(
+            t.device().is_cpu(),
```
Nit: couldn't we first check if the tensor is on CPU and only then, in case it isn't, look its device index up in the map?

lw reviewed (approved):
Looks good, thanks!

mrshenli committed:
Update on "Make pipeWrte and pipeRead noexcept"


After the previous device maps commits, `pipeWrite` might throw. In
this case, if we increment active calls before `pipeWrite` on the
caller, that active call won't be decremented properly when `pipeWrite`
throws. As a result, `shutdown` can silently timeout. I noticed this
as some tests take more than 60s to finish.

This commit extract the tensor device checking logic out of pipeWrite,
and make sure the error is thrown before the active call count is
incremented.

Differential Revision: [D24094803](https://our.internmc.facebook.com/intern/diff/D24094803)

[ghstack-poisoned]

mrshenli committed:
Update on "Make pipeWrte and pipeRead noexcept"


After the previous device maps commits, `pipeWrite` might throw. In
this case, if we increment active calls before `pipeWrite` on the
caller, that active call won't be decremented properly when `pipeWrite`
throws. As a result, `shutdown` can silently timeout. I noticed this
as some tests take more than 60s to finish.

This commit extract the tensor device checking logic out of pipeWrite,
and make sure the error is thrown before the active call count is
incremented.

Differential Revision: [D24094803](https://our.internmc.facebook.com/intern/diff/D24094803)

[ghstack-poisoned]

mrshenli committed:
Update on "Make pipeWrte and pipeRead noexcept"


After the previous device maps commits, `pipeWrite` might throw. In
this case, if we increment active calls before `pipeWrite` on the
caller, that active call won't be decremented properly when `pipeWrite`
throws. As a result, `shutdown` can silently timeout. I noticed this
as some tests take more than 60s to finish.

This commit extract the tensor device checking logic out of pipeWrite,
and make sure the error is thrown before the active call count is
incremented.

Differential Revision: [D24094803](https://our.internmc.facebook.com/intern/diff/D24094803)

[ghstack-poisoned]

mrshenli committed:
Update on "Make pipeWrte and pipeRead noexcept"


After the previous device maps commits, `pipeWrite` might throw. In
this case, if we increment active calls before `pipeWrite` on the
caller, that active call won't be decremented properly when `pipeWrite`
throws. As a result, `shutdown` can silently timeout. I noticed this
as some tests take more than 60s to finish.

This commit extract the tensor device checking logic out of pipeWrite,
and make sure the error is thrown before the active call count is
incremented.

Differential Revision: [D24094803](https://our.internmc.facebook.com/intern/diff/D24094803)

[ghstack-poisoned]

mrshenli committed:
Update on "Make pipeWrte and pipeRead noexcept"


After the previous device maps commits, `pipeWrite` might throw. In
this case, if we increment active calls before `pipeWrite` on the
caller, that active call won't be decremented properly when `pipeWrite`
throws. As a result, `shutdown` can silently timeout. I noticed this
as some tests take more than 60s to finish.

This commit extract the tensor device checking logic out of pipeWrite,
and make sure the error is thrown before the active call count is
incremented.

Differential Revision: [D24094803](https://our.internmc.facebook.com/intern/diff/D24094803)

[ghstack-poisoned]
